cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -170,9 +170,10 @@ Result<compute::Expression> FromProto(const substrait::Expression& expr,
               out = compute::field_ref(FieldRef(*out_ref, index));
             } else if (out->call() && out->call()->function_name == "struct_field") {
               // Nested StructFields on top of an arbitrary expression
-              std::static_pointer_cast<arrow::compute::StructFieldOptions>(
-                  out->call()->options)
-                  ->indices.push_back(index);
+              auto* field_options =
+                  checked_cast<compute::StructFieldOptions*>(out->call()->options.get());
+              field_options->field_ref =
+                  FieldRef(std::move(field_options->field_ref), index);
```

Comment:
@westonpace @vibhatha Can you look at the Substrait changes here?

Reply:
@pitrou Sure will take a look.

Reply:
@westonpace does the string fields matter here?

Reply:
in FromProto, I assume the indices will always be integers (since that's how substrait's field references work?)

Reply:
@jorisvandenbossche Ah, you might be right.

Reply:
Yes that's what I am referring to, Substrait doesn't consider names, only integers (position basically). 
But want to see @westonpace thoughts on this too.

Reply:
Yes, `FromProto` will always have integer indices (note that `index` comes from `ref->struct_field().field()` which is `int`).  So I don't see any harm here.  This looks good to me.